# Fake news detection

## Methods: TFIDF and PassiveAgressiveClassifier

<span style="color:blue"> **Dataset**</span>
:<br> * news.csv *:<br>
The first column identifies the news, the second and third are the title and text, and the fourth column has labels denoting whether the news is REAL or FAKE.

In [31]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.metrics import classification_report, accuracy_score

In [14]:
df= pd.read_csv('news.csv')

In [15]:
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [12]:
labels=df.label

In [13]:
labels

0       FAKE
1       FAKE
2       REAL
3       FAKE
4       REAL
        ... 
6330    REAL
6331    FAKE
6332    FAKE
6333    REAL
6334    REAL
Name: label, Length: 6335, dtype: object

## Split dataset to training and test sets

In [16]:
x_train,x_test,y_train,y_test = train_test_split(df['text'],labels, test_size=0.2, random_state=8)

In [19]:
x_test.shape

(1267,)

### Let’s initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7

In [35]:
#DataFlair - Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)
#DataFlair - Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

### Next, we’ll initialize a PassiveAggressiveClassifier. This is. We’ll fit this on tfidf_train and y_train.

In [32]:
#DataFlair - Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)
#DataFlair - Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')
print(f"Classification Report : \n\n{classification_report(y_test, y_pred)}")

Accuracy: 93.37%
Classification Report : 

              precision    recall  f1-score   support

        FAKE       0.93      0.94      0.93       636
        REAL       0.93      0.93      0.93       631

    accuracy                           0.93      1267
   macro avg       0.93      0.93      0.93      1267
weighted avg       0.93      0.93      0.93      1267



In [36]:
y_test.shape

(1267,)

### Calculate the accuracy and Classification report

In [22]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[598,  38],
       [ 43, 588]], dtype=int64)

<span style="color:red"> 598 true positives, 588 true negatives, 38 false positives, and 43 false negatives.</span>